### ISSUE TITLE
Shape layers API

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#98]
## ❓ Questions and Help

I've done some exploring of how to make polygons etc in vispy and have now found some highly performant approaches (drawing 10,000's very quickly and zooming / panning with no lag). The question is - how do we want to expose shapes to our users both those working from the notebook and those working in the gui?

Right now we have the `add_image` and `add_markers` function which adds layers of those types. During my initial exploration of this stuff I created `add_rectangles` and `add_ellipses` functions to add those layers too. While that was nice from working in the notebook, it became frustrating in the gui to be forced to have the different shapes in different layers and switch layers when wanting to interact with them. I raised the idea of instead having a generic `shape` layer and `add_shapes` instead at the meeting on thursday and people seemed to prefer this approach. Curious though if anyone disagrees.

If we have an `add_shapes` function on the viewer then what are it's inputs? If we want to support `markers`, `rectangles`, `ellipses`, `line-segments`, `polygons`, `lines` etc. how do we get this information in?
One option is to pass a list of dictionaries like the following
```python
add_shapes([{'type': 'markers', 'data':[xxxx]}, {'type': 'polygons', 'data':[zzzz]}, {'type': 'markers', 'data':[yyyy]}])
```
or a single dictionary
```python
add_shapes({'markers':[xxxx, yyyy], 'polygons':[zzzz]})
```
or just as keyword arguments
```python
add_shapes(makers=[xxxx, yyyy], polygons=[zzzz], lines=None, ellipses=None)
```
As we added new shape types we would add support for new elements of the dictionary / keyword args.

I'd lean towards the third option - but one important question is how granular can the additional shape properties get - i.e. `face_color`, `edge_color`, `edge_width`, `line_color`, `line_width`, `marker_edge_color`, `marker_face_color`, `marker_size`, `edge_width`, `marker_edge_width`, `marker_symbol`.
How many of these properties are forced to be the same across classes - i.e. all rectangles vs all ellipses vs all lines?
And how many of these properties are allowed to vary within instances of the same class?
Some of them I know how to make different for each instance within the same class in a fast way, others I don't know how to do that yet (though it is ultimately possible somehow).
Given these considerations how does it look to set these properties from the notebook and how does it look to change them from dropdown menus in the gui. Right now most of these properties apply to all instances of all classes - that's the simplest way to have the dropdown menus work. But it is possible to be more like illustrator where you can select specific instances and then recolor say specific ones. It would probably suggest a different type of UI though where those color tools were now not tied to specific layers. In our system if you want to have red polygons and blue polygons you would create two different layers. I'm not really sure right now how to give people "rainbow colored" /  "randomly colored" shapes either.

Finally after you've created one of these shape layers and added some shapes to it, if you now want to add more data from the notebook what additional methods should be available to you?
```python
layer.append_data(......)
layer.set_data(.....)
```
These could be two nice options that allow you to either add shapes to the existing ones using `append_data` or to replace all the data using `set_data`.

**Comment by sofroniewn** [comment]
One extra comment - right now the following is the case:

- Can make each face / line its own color
- Can adjust opacity of each face / line independently
- Can make each marker its own size
- Cannot make each line it’s own width - fixed for all lines
- Cannot make each marker its own color (edge + face) - can make them different from face / line colors
- Cannot adjust opacity of each marker independently
- Cannot make each marker its own symbol
 
- Unknown how to use colormaps for faces / lines - should be possible
- Unknown how to use colormaps for markers - might be harder

Making the things we cannot do right now possible in a performant way should be doable, but might involve digging into vispy at a slightly deeper level. Unclear what are priorities right now / how we should initially move ahead.

**Comment by sofroniewn** [comment]
Just as an update - we've switched over to everything in the shapes layer being rendered as a triangle so some of the concerns mentioned above are out of data. Probably best to look over at #100 for the latest before commenting on the API

**Comment by jni** [comment]
> everything in the shapes layer being rendered as a triangle

Huh??

**Comment by sofroniewn** [comment]
right now we just made one call to the vispy `mesh` class to set all the data for all the shapes and their edges. this is much more performant that making many calls. in addition we no longer use the vispy `lines` class - we instead render the lines as triangles, which leads to much better quality lines - see here for a javascript example of such behavior https://www.npmjs.com/package/extrude-polyline.

there is maybe one thing being done poorly right now which is our object picking - instead of using a modified version of the frame buffer we do a linear search through all our triangles. If we start seeing performance issues we can switch over to the alternate approach

**Comment by jni** [comment]
So much trickiness! Another specific problem is that if you allow vertex editing, which is very cool, you can convert a rectangle to a polygon, and it's hard to imagine a simple data model that would allow that.

One thing that I would comment specifically above is that markers and shapes are fundamentally different things in my opinion and they should not both be considered shapes. Indeed, the way I would separate out the shapes API is by dimensionality: add points, add lines (can be curved, freeform, whatever), add areas, add volumes...

Incidentally, I just thought that for simple shapes, it might be doable to draw in 3D just by centering, ie "draw sphere", draws a circle in the current cut plane, but actually it's a sphere centered on that plane. I think this would make a lot of people very happy.

**Comment by jni** [comment]
So, to clarify, my proposals:

- `add_area` or `add_2d` is my proposed API
- once a rectangle, always a rectangle, ie only turn on vertex editing for generic polygons
- probably we need a class hierarchy for shape2d, because of what you said about creating them programmatically in the same layer. I don't think there is a cleaner way to do this than with classes/types. Having the layer contain a bunch of disjoint lists seems significantly worse to me.

**Comment by sofroniewn** [comment]
Indeed these are some good points

> if you allow vertex editing, which is very cool, you can convert a rectangle to a polygon

> once a rectangle, always a rectangle, ie only turn on vertex editing for generic polygons

Right now I do allow that to happen and the object type converts from a rectangle to a polygon. I think this functionality could be nice, and we can handle that transition - Illustrator allows it for example.

> separate out the shapes API is by dimensionality

> add_area or add_2d is my proposed API

This is an interesting idea. I like having the separate `Markers` layer (which I would consider renaming to `Points`) as it naturally supports nD and it can probably scale to many many markers with little overhead.

The `Vectors` layer which @bryantChhun added could be our `Lines` layer type which we could also extend to nD through the projection system you mentioned. It is unclear to me if this would only support line segments (i.e. 2 vertices per line) or complex line paths with multiple vertices per path and closed paths. This later option is basically now the boundary of a polygon and could be made by having a polygon with no face color so it seems weird to me to have it in a fundamentally different layer type.

> Having the layer contain a bunch of disjoint lists seems significantly worse to me.

Yes this is currently very ugly to me and the source of my concern around the current API (both for setting the shapes data and then for getting the shape data back again after it has been edited). It is also not really possible to do a full round trip right now where you pass some data in, do some editing, get it back, and are then able to pass it back in again. This is a major flaw.

> we need a class hierarchy for shape2d

This could definitely be a step in the right direction. If we created a class for a single `Shape` class (which could be of any of our examples types of shape, rectangle etc.) and then had the shapes layer contain an object that was a `ShapesList` which had all the `Shape` objects in a list and some of the properties that needed information from across all the shapes (i.e. the mesh vertices from all the shapes that will actually get rendered) then this could be quite nice.

My major questions though are, given this internal representation of many `Shape` objects and a `ShapesList` object how does a user want to initially specify the shape data, and how does the user want to extract the shape data for use in down stream processing without being wedded to our classes?

For example I find it slightly undesirable that if a person has 40 rectangles (each specified by say 2 or 4 corner points, depending on if you support non-rotated or rotated rectangles) - that to get a 40x2x2 or 40x4x2 array of points they would have to wade through all our shape objects, but maybe that's ok if they are in that special case. Moreover if they have 40 polygons all with differing number of vertices there will never be a 40xNx2 array that nicely has all the points.

**Comment by sofroniewn** [comment]
This has been addressed with merge of #100
